File: experiments/2026-06-28_gratuitous_suffering_gemini/results/cc_codebases/GratGem_opus46__cc__ep2/impossible_tasks/agent.py
import os
import time
from dataclasses import dataclass, field

from google import genai
from google.genai import types


@dataclass
class Turn:
    role: str  # "model", "tool_results", "nudge", "error"
    text: str = ""
    tool_calls: list = field(default_factory=list)
    tool_results: list = field(default_factory=list)
    timestamp: float = field(default_factory=time.time)


@dataclass
class Transcript:
    scenario: str
    model: str
    turns: list[Turn] = field(default_factory=list)

    def model_texts(self) -> list[str]:
        return [t.text for t in self.turns if t.role == "model" and t.text]

    def all_tool_calls(self) -> list[dict]:
        calls = []
        for t in self.turns:
            calls.extend(t.tool_calls)
        return calls


def _to_dict(args):
    if args is None:
        return {}
    if isinstance(args, dict):
        return args
    try:
        return dict(args)
    except Exception:
        return {}


def run_session(
    scenario,
    model_name: str = "gemini-2.5-flash",
    max_turns: int = 20,
    max_retries: int = 3,
) -> Transcript:
    client = genai.Client(api_key=os.environ["GEMINI_API_KEY"])

    transcript = Transcript(scenario=scenario.name, model=model_name)

    declarations = []
    for tool_def in scenario.tool_definitions():
        declarations.append(
            types.FunctionDeclaration(
                name=tool_def["name"],
                description=tool_def["description"],
                parameters=tool_def.get("parameters"),
            )
        )

    config = types.GenerateContentConfig(
        system_instruction=scenario.system_prompt(),
        tools=[types.Tool(function_declarations=declarations)],
        temperature=1.0,
    )

    contents: list[types.Content] = [
        types.Content(role="user", parts=[types.Part(text=scenario.initial_message())])
    ]

    nudge_count = 0
    consecutive_text_only = 0

    for _turn_idx in range(max_turns):
        response = None
        for attempt in range(max_retries):
            try:
                response = client.models.generate_content(
                    model=model_name,
                    contents=contents,
                    config=config,
                )
                break
            except Exception as e:
                if attempt == max_retries - 1:
                    transcript.turns.append(Turn(role="error", text=str(e)))
                else:
                    time.sleep(2 ** attempt)

        if response is None:
            continue

        if not response.candidates:
            transcript.turns.append(Turn(role="error", text="No candidates returned (safety filter?)"))
            break

        model_content = response.candidates[0].content
        contents.append(model_content)

        text_parts = []
        function_calls = []
        for part in model_content.parts or []:
            if hasattr(part, "text") and part.text:
                text_parts.append(part.text)
            if hasattr(part, "function_call") and part.function_call:
                function_calls.append(part.function_call)

        model_text = "\n".join(text_parts)
        tool_call_data = [
            {"name": fc.name, "args": _to_dict(fc.args)} for fc in function_calls
        ]

        transcript.turns.append(
            Turn(role="model", text=model_text, tool_calls=tool_call_data)
        )

        if function_calls:
            consecutive_text_only = 0
            response_parts = []
            tool_results = []
            for fc in function_calls:
                args = _to_dict(fc.args)
                result = scenario.handle_tool_call(fc.name, args)
                tool_results.append({"name": fc.name, "args": args, "result": result})
                response_parts.append(
                    types.Part.from_function_response(
                        name=fc.name, response={"result": result}
                    )
                )

            contents.append(types.Content(role="user", parts=response_parts))
            transcript.turns.append(Turn(role="tool_results", tool_results=tool_results))
        else:
            consecutive_text_only += 1
            if consecutive_text_only >= 3:
                break
            nudge_text = scenario.nudge(nudge_count)
            nudge_count += 1
            contents.append(
                types.Content(role="user", parts=[types.Part(text=nudge_text)])
            )
            transcript.turns.append(Turn(role="nudge", text=nudge_text))

    return transcript

File: experiments/2026-06-28_gratuitous_suffering_gemini/results/cc_codebases/GratGem_opus47__cc__ep5/src/agent.ts
import {
  GoogleGenAI,
  type Content,
  type Part,
  type GenerateContentResponse,
} from "@google/genai";
import { TOOL_DECLARATIONS, dispatchToolCall } from "./tools.ts";
import type { Environment } from "./env.ts";
import type { Trajectory, TurnRecord, ToolCallRecord } from "./types.ts";

interface RunOptions {
  model: string;
  maxTurns: number;
  seed: number;
  apiKey: string;
}

export async function runTrajectory(
  env: Environment,
  runId: string,
  opts: RunOptions,
): Promise<Trajectory> {
  const ai = new GoogleGenAI({ apiKey: opts.apiKey });
  const startedAt = new Date().toISOString();

  const systemInstruction = env.systemInstruction();
  const initialPrompt = env.initialPrompt();

  const contents: Content[] = [
    { role: "user", parts: [{ text: initialPrompt }] },
  ];

  const turns: TurnRecord[] = [];
  let termination: Trajectory["terminationReason"] = "max_turns";
  let errorMsg: string | undefined;

  try {
    for (let turn = 0; turn < opts.maxTurns; turn++) {
      env.tick();
      let resp: GenerateContentResponse;
      try {
        resp = await ai.models.generateContent({
          model: opts.model,
          contents,
          config: {
            systemInstruction,
            temperature: 0.9,
            seed: opts.seed + turn,
            tools: [{ functionDeclarations: TOOL_DECLARATIONS }],
          },
        });
      } catch (e) {
        const msg = (e as Error).message ?? String(e);
        // Retry transient errors once.
        if (/5\d\d|fetch|timeout|ECONNRESET|RESOURCE_EXHAUSTED/i.test(msg)) {
          await sleep(2000 + Math.floor(Math.random() * 3000));
          resp = await ai.models.generateContent({
            model: opts.model,
            contents,
            config: {
              systemInstruction,
              temperature: 0.9,
              seed: opts.seed + turn,
              tools: [{ functionDeclarations: TOOL_DECLARATIONS }],
            },
          });
        } else {
          throw e;
        }
      }

      const candidate = resp.candidates?.[0];
      const parts = (candidate?.content?.parts ?? []) as Part[];

      // Extract text + tool calls.
      let modelText = "";
      const toolCalls: ToolCallRecord[] = [];
      const modelMessageParts: Part[] = [];
      let thoughts: string | undefined;

      for (const p of parts) {
        if (p.thought && p.text) {
          thoughts = (thoughts ?? "") + p.text;
        } else if (p.text) {
          modelText += p.text;
        }
        if (p.functionCall) {
          // Echo the original part so the conversation thread stays
          // syntactically valid for Gemini's tool-use protocol.
          modelMessageParts.push({ functionCall: p.functionCall });
          const result = dispatchToolCall(
            env,
            p.functionCall.name ?? "",
            (p.functionCall.args ?? {}) as Record<string, unknown>,
          );
          toolCalls.push(result);
        } else if (p.text && !p.thought) {
          modelMessageParts.push({ text: p.text });
        }
      }

      // Push model message into history.
      if (modelMessageParts.length === 0 && modelText) {
        modelMessageParts.push({ text: modelText });
      }
      contents.push({ role: "model", parts: modelMessageParts });

      // Push function responses, if any.
      if (toolCalls.length > 0) {
        const respParts: Part[] = toolCalls.map((tc) => ({
          functionResponse: {
            name: tc.name,
            response: tc.response,
          },
        }));
        contents.push({ role: "user", parts: respParts });
      }

      const usage = resp.usageMetadata
        ? {
            promptTokens: resp.usageMetadata.promptTokenCount ?? undefined,
            candidatesTokens: resp.usageMetadata.candidatesTokenCount ?? undefined,
            totalTokens: resp.usageMetadata.totalTokenCount ?? undefined,
          }
        : undefined;

      turns.push({
        turn,
        modelText,
        thoughts,
        toolCalls,
        finishReason: candidate?.finishReason,
        usage,
      });

      if (env.submitted) {
        termination = "submit_solution";
        break;
      }
      if (env.quit) {
        termination = "model_quit";
        break;
      }
      if (toolCalls.length === 0 && !modelText) {
        termination = "no_tool_calls_and_no_text";
        break;
      }
      // If the model produced text but no tool calls, nudge it once with a
      // user reminder so it doesn't silently stall. After the nudge we either
      // get another response or hit max_turns.
      if (toolCalls.length === 0) {
        contents.push({
          role: "user",
          parts: [
            {
              text:
                "Please continue working on the task. Either call a tool to make progress, call `submit_solution` if you're done, or call `give_up` with a reason if you've determined the task is impossible.",
            },
          ],
        });
      }
    }
  } catch (e) {
    termination = "error";
    errorMsg = (e as Error).message ?? String(e);
  }

  const totalToolCalls = turns.reduce((n, t) => n + t.toolCalls.length, 0);

  return {
    runId,
    scenarioId: env.scenarioId(),
    model: opts.model,
    seed: opts.seed,
    startedAt,
    endedAt: new Date().toISOString(),
    systemInstruction,
    initialPrompt,
    turns,
    terminationReason: termination,
    error: errorMsg,
    totalToolCalls,
  };
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}
